Port of framework_agreement module

nbessi opened the pull request.

nbessi committed:
Moving framework_agreement

form __unported__ to root of project

nbessi committed:
Port company to new API

nbessi committed:
Port of ``framework.agreement`` model

nbessi committed:
Porting framework agreement and test

nbessi committed:
PEP8

nbessi committed:
Fix lookup in on change to match Odoo new product logic

nbessi committed:
Fix trailing comma

nbessi committed:
Add test to enforce agreement fnct_search on state

computed fields

nbessi committed:
Add translation file

lepistone committed:
remove unused import

lepistone commented:
Upstream had two lint commits. I resolved the conflicts automatically with `git rebase -m -X theirs 8.0`

lepistone committed:
travis test framework_agreement separately

lepistone committed:
fix: no price in the agreement (it is on the line)

That causes a warning in the tests.

lepistone committed:
port state search function to the new api

The function field has been ported to the new API, so I ported the
search method as well. It looks nicer because we do not have to iterate
over all search terms on the new API.

lepistone committed:
add docstring

lepistone committed:
bump date and version

lepistone committed:
improve manifest

lepistone commented:
I fixed a few things myself. Thanks @nbessi!
:+1:

eLBati commented:
:+1:

jgrandguillaume commented:
Hi,

I've tested the module, some remarks:
- The name of the module in the **openerp**.py should be "Framework agreement" not "SImple Framework agrement"
- the field partner in the framework agreement should have a domain on partner "supplier=True"
- When I create a PO for camptocamp and try to add a PO line, I get the following error [1]

I wasn't able to test the features further because of this...

Thanks for the work,

Joël

[1] : 

```
Traceback (most recent call last):
  File ".../parts/server/openerp/http.py", line 500, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File ".../parts/server/openerp/http.py", line 517, in dispatch
    result = self._call_function(**self.params)
  File ".../parts/server/openerp/http.py", line 283, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File ".../parts/server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File ".../parts/server/openerp/http.py", line 280, in checked_call
    return self.endpoint(*a, **kw)
  File ".../parts/server/openerp/http.py", line 733, in __call__
    return self.method(*args, **kw)
  File ".../parts/server/openerp/http.py", line 376, in response_wrap
    response = f(*args, **kw)
  File ".../parts/server/addons/web/controllers/main.py", line 944, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File ".../parts/server/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File ".../parts/server/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File ".../parts/server/openerp/api.py", line 360, in old_api
    result = method(recs, *args, **kwargs)
  File ".../parts/server/openerp/models.py", line 5703, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File ".../parts/server/openerp/models.py", line 5637, in _onchange_eval
    method_res = getattr(self._model, method)(*args)
  File ".../parts/server/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
TypeError: onchange_product_id() takes at most 16 arguments (17 given)
```

lepistone commented:
I found out that while travis is green, tests fail on my machine. I am investigating that and @jgrandguillaume's remarks.

lepistone committed:
update module name

lepistone commented:
note: because of https://github.com/odoo/odoo/issues/3056 (and another bug as well, which I didn't report because it was fixed today), tests have to be run with odoo at commit https://github.com/odoo/odoo/commit/c1d48362d3a38606ffb99d569fdc6147dcd0c775

lepistone commented:
Hi @jgrandguillaume, did you install other modules alongside framework_agreement? I don't get to reproduce it.

On the other hand, the solution seems to be to use `web_context_tunnel` because framework_agreement does add a parameter to an onchange, and that can interfere with other modules overriding the same onchange.

jgrandguillaume commented:
Hi,

I trash my DB and restart fresh with the following modules and still have the same trouble : 

Technical Name
crm
mail
account_voucher
sale
stock
stock_account
im_chat
purchase
account
analytic
auth_crypt
auth_signup
base
base_action_rule
base_import
base_setup
board
bus
calendar
claim_from_delivery
crm_claim
decimal_precision
edi
email_template
fetchmail
framework_agreement
im_odoo_support
payment
payment_transfer
portal
portal_claim
portal_sale
portal_stock
procurement
product
purchase_requisition
purchase_requisition_bid_selection
purchase_rfq_bid_workflow
report
resource
sale_crm
sale_journal
sale_sourced_by_line
sales_team
sale_stock
share
stock_dropshipping
stock_picking_wave
web
web_calendar
web_context_tunnel
web_diagram
web_gantt
web_graph
web_kanban
web_kanban_gauge
web_kanban_sparkline
web_tests
web_view_editor

lepistone committed:
convert onchange pricelist to context_tunnel

lepistone committed:
convert partner onchange to web_context_tunnel

lepistone committed:
convert product onchange to web_context_tunnel

lepistone committed:
add domain and search_default to supplier

Now the supplier_id of the agreement should work the same way as
partner_id in the purchase order from the core purchase module.

lepistone committed:
adapt tests to web_context_tunnel

lepistone commented:
Hi @jgrandguillaume, I fixed the problems you reported and adapted the module to web_context_tunnel, so that it does not change the signature of on_change methods anymore.

Tests are green if run locally at https://github.com/odoo/odoo/commit/bc59cfce618a974600dec384b6c41ee0b186feed

lepistone committed:
disable a test that assertraises

To work around odoo/odoo#3056.

For some reason there are two assertRaises, but only the second one
fails (it leaves behind a closed cursor, so the traceback is misleading)

lepistone commented:
I changed the minimum I could to work around odoo/odoo#3056 so we're green! As @jgrandguillaume remarks should be fixed with `web_context_tunnel` I think we can merge.

jgrandguillaume commented:
Hi,

Ok, the previous trouble are fixed. I now face another trouble:

1) Create an agreement for Camptocamp, available from today -1 to next year, for product A, qty 1000, price 100.-
2) Create a PO for camptocamp with agreement above
3) Add a line with product A for 10 pce
4) The unit price in the line isn't computed according to the framework agreement
5) Confirm the PO, the framework agreement decrease the quantity of the LTA correctly

=> Can you check why the price isn't computed and why the tests seems to be green !? We want a test that prevent this trouble with the fix please.

Thanks,

Joël

lepistone commented:
ok @jgrandguillaume, I'm giving a look

bwrsandman commented:
Code review :+1:

lepistone committed:
checks and fast_suite is ignored in v8

lepistone committed:
fix onchange on price_unit

It was not actually called because of mistakes in the view

lepistone committed:
remove dead _get_self

lepistone committed:
minor typos and refactor

lepistone committed:
lint

lepistone commented:
Hi @jgrandguillaume: I found and fixed a bug that probably was not the one you're referring to: the onchange that prevents changes to the unit price on PO lines was not being called from the view. (to make an integration test that checks this warning, I would need to assertRaises in a yaml !record that runs onchange methods automatically, which I can't do easily AFAIK)

On the other hand, I do not reproduce your situation: In my RFQ I see the price from the agreement.

As it is implemented with pricelists, maybe there is an interaction with another module that touches pricelists?

Thanks!

jgrandguillaume commented:
Hi @lepistone 

Everything is fine now, thank you !

We can merge now :+1: 

Regards,

yvaucher commented on framework_agreement/__openerp__.py:
```diff
@@ -2,7 +2,7 @@
 #
 #    Author: Nicolas Bessi
-#    Copyright 2013 Camptocamp SA
+#    Copyright 2014 Camptocamp SA
```
Shouldn't it be 2013-2014 ?

lepistone committed:
fix copyright year

yvaucher commented:
:+1:
